File: openmp-wg-2015/omp_ndot_local.f90
!//
!//
!//

real*8 function ndot(N,x,y)
   integer :: N
   real*8 :: x(N),y(N),tot,mytot
   tot=0.0
   mytoto = 0.0
!$OMP parallel default(none) shared(tot,n,x,y) private(mytot)
!$OMP do
   do i=1,N
      mytot = mytot + (x(i) * y(i))
   end do
!$OMP end do
!$OMP atomic
   tot = tot + mytot
!$OMP END parallel
   ndot = tot
end function

program omp_ndot_loca

   integer,parameter :: N = 1e7
   real*8 :: x(N),y(N),ans,dot,ndot

   do i=1,N 
      x(i) = i-1.
      y(i) = i-1.
   end do

   ans=(N-1.)*N*(2.0*N-1.)/6.0;

   call tick(it0,rate);
   dot = ndot(N,x,y)
   call tock(it0,rate)
   write(*,"(A,ES12.4,A,ES12.4,A,I10)") "Dot product: ",dot," (vs ",ans,") for n = ",n
end program

File: openmp-wg-2015/omp_ndot_local.c
// ndot.c
//
// Inner product example. Serial version.
//
//
#include <stdio.h>
#include "ticktock.h"

double ndot(int n, double *x, double *y)
{
   double tot=0;
   #pragma omp parallel default(none) shared(tot,n,x,y)
   {
   double mytot = 0.;
   int i;
   #pragma omp for
   for (i=0; i<n; i++)
      mytot += x[i] * y[i];
   #pragma omp atomic
      tot += mytot;
   }
   return tot;
}

int main() 
{
   int i,n = 1e7;
   double*  x = malloc(sizeof(double) * n);
   double*  y = malloc(sizeof(double) * n);

   for (i=0; i<n; i++) 
      x[i]=y[i]=i;

   double ans=(n-1.)*n*(2.0*n-1.)/6.0;

   tick_tock tt;
   tick(&tt);
   double dot = ndot(n,x,y);
   printf("Dot product: %8.4e (vs %8.4e) for n = %d\n",dot,ans,n);
   tock(&tt);

   free(x); 
   free(y);
}
